storage/ndb/src/kernel/blocks/dbtup/AggInterpreter.cpp

```diff
@@ -0,0 +1,2138 @@
+   */
+  if (n_gb_cols_) {
+#ifdef MOZ_AGG_MALLOC
+    assert(n_gb_cols_ <= MAX_AGG_N_GROUPBY_COLS);
+    gb_cols_ = gb_cols_buf_;
+#else
+    gb_cols_ = new uint32_t[n_gb_cols_];
+#endif // MOZ_AGG_MALLOC
+
+    uint32_t i = 0;
+    while (i < n_gb_cols_ && cur_pos_ < prog_len_) {
+      gb_cols_[i++] = prog_[cur_pos_++];
+    }
+#ifdef MOZ_AGG_MALLOC
+    gb_map_ = &gb_map_buf_;
+#else
+    gb_map_ = new std::map<GBHashEntry, GBHashEntry, GBHashEntryCmp>;
+#endif // MOZ_AGG_MALLOC
+  }
+
+  /*
+   * 4. Reset all aggregation results
+   */
+  if (n_agg_results_) {
+#ifdef MOZ_AGG_MALLOC
+    assert(n_agg_results_ <= MAX_AGG_N_RESULTS);
+    agg_results_ = agg_results_buf_;
+#else
+    agg_results_ = new AggResItem[n_agg_results_];
+#endif // MOZ_AGG_MALLOC
+    uint32_t i = 0;
+    while (i < n_agg_results_) {
+      agg_results_[i].type = NDB_TYPE_UNDEFINED;
+      agg_results_[i].value.val_int64 = 0;
+      agg_results_[i].is_unsigned = false;
+      agg_results_[i].is_null = true;
+      i++;
+    }
+  }
+
+  inited_ = true;
+  agg_prog_start_pos_ = cur_pos_;
+  memset(registers_, 0, sizeof(registers_));
+
+  return true;
+}
+
+static bool TypeSupported(DataType type) {
+  switch (type) {
+    case NDB_TYPE_TINYINT:
+    case NDB_TYPE_SMALLINT:
+    case NDB_TYPE_MEDIUMINT:
+    case NDB_TYPE_INT:
+    case NDB_TYPE_BIGINT:
+
+    case NDB_TYPE_TINYUNSIGNED:
+    case NDB_TYPE_SMALLUNSIGNED:
+    case NDB_TYPE_MEDIUMUNSIGNED:
+    case NDB_TYPE_UNSIGNED:
+    case NDB_TYPE_BIGUNSIGNED:
+
+    case NDB_TYPE_FLOAT:
+    case NDB_TYPE_DOUBLE:
+
+    case NDB_TYPE_DECIMAL:
+    case NDB_TYPE_DECIMALUNSIGNED:
+      return true;
+    default:
+      return false;
+  }
+  return false;
+}
+
+static bool IsUnsigned(DataType type) {
+  switch (type) {
+    case NDB_TYPE_TINYUNSIGNED:
+    case NDB_TYPE_SMALLUNSIGNED:
+    case NDB_TYPE_MEDIUMUNSIGNED:
+    case NDB_TYPE_UNSIGNED:
+    case NDB_TYPE_BIGUNSIGNED:
+    case NDB_TYPE_DECIMALUNSIGNED:
+      return true;
+    default:
+      return false;
+  }
+  return false;
+}
+
+static DataType AlignedType(DataType type) {
+  switch (type) {
+    case NDB_TYPE_TINYINT:
+    case NDB_TYPE_SMALLINT:
+    case NDB_TYPE_MEDIUMINT:
+    case NDB_TYPE_INT:
+    case NDB_TYPE_BIGINT:
+
+    case NDB_TYPE_TINYUNSIGNED:
+    case NDB_TYPE_SMALLUNSIGNED:
+    case NDB_TYPE_MEDIUMUNSIGNED:
+    case NDB_TYPE_UNSIGNED:
+    case NDB_TYPE_BIGUNSIGNED:
+
+      return NDB_TYPE_BIGINT;
+    case NDB_TYPE_FLOAT:
+    case NDB_TYPE_DOUBLE:
+      return NDB_TYPE_DOUBLE;
+    /*
+     * TODO (Zhao)
+     * Moz
+     * Temporary solultion
+     * Currently regard Decimal as a undefined,
+     * then decide it as BIGINT/BIGUNSIGNED/DOUBLE in LoadColumn
+     * dynamically.
+     */
+    case NDB_TYPE_DECIMAL:
+    case NDB_TYPE_DECIMALUNSIGNED:
+      return NDB_TYPE_UNDEFINED;
+    default:
+      assert(0);
+  }
+  return NDB_TYPE_UNDEFINED;
+}
+
+static bool TestIfSumOverflowsUint64(uint64_t arg1, uint64_t arg2) {
```

Comment:
Move this function to place where it can be reused for other parts of RonDB data node (e.g. in the other interpreter).